src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5287,12 +5287,16 @@ protected synchronized void drain(boolean isFinalShutdown) throws IOException, I
             CommitLog.instance.shutdownBlocking();
 
             // wait for miscellaneous tasks like sstable and commitlog segment deletion
-            ScheduledExecutors.nonPeriodicTasks.shutdown();
-            if (!ScheduledExecutors.nonPeriodicTasks.awaitTermination(1, MINUTES))
-                logger.warn("Unable to terminate non-periodic tasks within 1 minute.");
-
             ColumnFamilyStore.shutdownPostFlushExecutor();
-            setMode(Mode.DRAINED, !isFinalShutdown);
+
+            try
+            {
+                ScheduledExecutors.shutdownNowAndWait(1, MINUTES);
+            }
+            finally
+            {
+                setMode(Mode.DRAINED, !isFinalShutdown);
```

Comment:
Interesting when a node should decide it is `DRAINED` if there's an exception. You can't attempt draining again if an exception is thrown after the collection of mutation stages. I'm guessing you structured it like this so you would retain behavior on timeout from the scheduled executor shutdown as before.

Reply:
Exactly my reasoning behind that. We do not have idempotent drain right now. This is quite sad to be honest. We can not "re-drain". So this is our only chance to be drained even it throws.